File: 00xx/000x/0009-PalindromeNumber/Solution.java
/*  Solution: Start at both ends and go inward
    Time complexity: O(n/2 + 1)
    Space complexity: O(1)
 */

import java.lang.Math;

class Solution {
    public boolean isPalindrome(int x) {
        if (x < 0) { return false; }
        int digits = 0;
        int mostSigDig, leastSigDig;
        while (x >= Math.pow(10, digits)) {
            digits++;
        }
        while (digits > 0) {
            mostSigDig = (int) ( x / Math.pow(10, --digits) );
            x -= mostSigDig * Math.pow(10, digits);
            if (digits > 0) {
                leastSigDig = x % 10;
                x /= 10;
                digits--;
            }
            else {
                leastSigDig = mostSigDig;
            }
            if (mostSigDig != leastSigDig) {
                return false;
            }
        }
        return true;
    }
}

File: 00xx/000x/0009-PalindromeNumber/Solution.py
# Solution: Log / Modulo (no string conversion)
# Time Complexity:  O(log(n)) (n is x)
# Space Complexity: O(1)

import math

class Solution:
  def isPalindrome(self, x: int) -> bool:
    if x < 0:
      return False
    elif x == 0:
      return True
    d = int(math.log10(x)) + 1
    while d > 0:
      if int(x / math.pow(10, d-1)) != x % 10:
        return False
      x %= math.pow(10, d-1)
      x = int(x/10)
      d -= 2
    return True
